src/sdk/cpp/lib/include/protocol_gateway.hpp

```diff
@@ -141,27 +276,103 @@ class ProtocolGateway {
     static const SubscribeOptions DefaultSubscribeOptions;
 
    public:
+    /**
+     * @brief Constructs a new ProtocolGateway.
+     *
+     * @param config A JSON representation of the configuration.
+     * @param display_name A name forwarded to each Adapter.
+     * @param platform_proto_only Controls whether to load only platform protocol Adapters.
+     */
     ProtocolGateway(const json& config, const std::string& display_name = "", bool platform_proto_only = false);
 
     virtual ~ProtocolGateway() = default;
 
+    /**
+     * @brief Publish a message to the Adapters.
+     * ProtocolGateway::Start() must be called before
+     * ProtocolGateway::Publish().
+     *
+     * @param topic The topic to publish to.
+     * @param msg The message to publish.
+     * @param opts Options to pass along to the Adapters.
+     */
     virtual void Publish(const std::string& topic, const std::string& msg, const PublishOptions& options = DefaultPublishOptions);
 
+    /**
+     * @brief Subscribe to messages from the Adapters.
+     * ProtocolGateway::Start() must be called before
+     * ProtocolGateway::Subscribe().
+     *
+     * @param topic The topic to subscribe to.
+     * @param on_msg_func_ptr A callback function to call when a message
+     * matching the subcription arrives.
+     * @param opts Options to pass along to the Adapters.
+     */
     virtual void Subscribe(const std::string& topic, on_msg_func_ptr on_msg, const SubscribeOptions& opts = DefaultSubscribeOptions);
 
+    /**
+     * @brief Subscribe to shared messages from the Adapters.
+     * ProtocolGateway::Start() must be called before
+     * ProtocolGateway::SubscribeShared().
+     *
+     * @param topic The topic to subscribe to.
+     * @param on_msg_func_ptr A callback function to call when a message
+     * matching the subcription arrives.
+     * @param opts Options to pass along to the Adapters.
+     */
     virtual void SubscribeShared(const std::string& group, const std::string& topic, on_msg_func_ptr on_msg, const SubscribeOptions& opts = DefaultSubscribeOptions);
 
+    /**
+     * @brief Initialize the ProtocolGateway by loading all the Adapters listed
+     * in the configuration. Throws ProtocolGatewayException if any of the
+     * Adapters fail to load.
+     */
     virtual void Initialize();
+
+    /**
+     * @brief Start the ProtocolGateway. ProtocolGateway::Initialize() must be
+     * called before ProtocolGateway::Start().
+     */
     virtual void Start();
+
+    /**
+     * @brief Stop the ProtocolGateway.  ProtocolGateway::Start() must be
+     * called before ProtocolGateway::Stop().
+     */
     virtual void Stop();
 
+    /**
+     * @brief Add an adapter to the set of adapters.
+     *
+     * @param adatper The new Adapter.
+     *
+     * @return bool
+     */
+    virtual bool Add(std::shared_ptr<Adapter> adapter);
```

Comment:
Has this method to be public?

If yes: Is it that simply to use the add --> Does it has side-effects or will it lead to an undefined behavior adding a adapter after initialization? Has to be re-initialized?

Reply:
It has to be accessible to the tests. I'll make it protected and then at least it's not accessible to without subclassing.